src/enclosure.c

```diff
@@ -171,3 +184,146 @@ int enclosure_open(const struct enclosure_device *enclosure)
 
 	return fd;
 }
+
+static struct enclosure_device *find_enclosure(char *enclosure_id)
+{
+	struct enclosure_device *enclosure_device;
+	list_for_each(sysfs_get_enclosure_devices(), enclosure_device) {
+		if (strncmp(enclosure_device->dev_path, enclosure_id, PATH_MAX) == 0) {
+			return enclosure_device;
+		}
+	}
+	return NULL;
+}
+
+static struct ses_slot *find_enclosure_slot_by_index(struct enclosure_device *encl, int index)
+{
+	for (int i = 0; i < encl->slots_count; i++) {
+		if (encl->slots[i].index == index) {
+			return &encl->slots[i];
+		}
+	}
+	return NULL;
+}
+
+static status_t enclosure_get_slot_by_slot_num(char *slot_num, struct slot_response *slot_res)
+{
+	char tmp_enclosure_id[PATH_MAX];
+	char *enclosure_id;
+	char *index_str;
+	int index = -1;
+	struct enclosure_device *encl = NULL;
+	struct block_device *block_device = NULL;
+	struct ses_slot *s_slot = NULL;
+
+	snprintf(tmp_enclosure_id, PATH_MAX, "%s", slot_num);
+	snprintf(slot_res->slot, PATH_MAX, "%s", slot_num);
+
+	enclosure_id = dirname(tmp_enclosure_id);	// POSIX dirname modifies input
+	index_str = basename(slot_num);				// GNU basename does not
+
+	if (str_toi(&index, index_str, NULL, 10)) {
+		log_error("SCSI: Invalid slot identifier index %s\n", index_str);
+		return STATUS_INVALID_PATH;
+	}
+
+	encl = find_enclosure(enclosure_id);
+	if (!encl) {
+		log_error("SCSI: Invalid enclosure ='%s'\n", enclosure_id);
+		return STATUS_NULL_POINTER;
+	}
+
+	s_slot = find_enclosure_slot_by_index(encl, index);
+	if (!s_slot) {
+		log_error("SCSI: Unable to locate slot in enclosure %d\n", index);
+		return STATUS_NULL_POINTER;
+	}
+
+	slot_res->state = s_slot->ibpi_status;
+	block_device = locate_block_by_sas_addr(s_slot->sas_addr);
+
+	// Not having a block device for a slot is OK
+	if (block_device) {
+		snprintf(slot_res->device, PATH_MAX, "/dev/%s", basename(block_device->sysfs_path));
+	}
+
+	return STATUS_SUCCESS;
+}
+
+static status_t enclosure_get_slot_by_device(char *device, struct slot_response *slot_res)
+{
+	char slot_id[PATH_MAX];
+	struct block_device *block_device = get_block_device_from_sysfs_path(basename(device));
+
+	if (!block_device) {
+		log_error("SCSI: Device node not found %s\n", device);
+		return STATUS_INVALID_PATH;
+	}
+
+	if (!block_device->enclosure) {
+		log_error("SCSI: Not a SCSI ses device %n\n", device);
+		return STATUS_INVALID_PATH;
+	}
+
+	snprintf(slot_id, PATH_MAX, "%s/%d", block_device->enclosure->dev_path, block_device->encl_index);
+	return enclosure_get_slot_by_slot_num(slot_id, slot_res);
+}
+
+status_t enclosure_get_slot(char *device, char *slot_num, struct slot_response *slot_res)
```

Comment:
OK, I believe I've refactored the code guided by your suggestion, but it does vary with the following:

*  I did leave an additional function `enclosure_get_slot_by_slot_num` instead of placing its implementation in `enclosure_get_slot`
* Added the retrieval of the device node to the function `_enclosure_get_slot`
* Created a new function to parse the slot number which is called `pars_slot_num` as we had 2 places with the same code to do this, once in each of the get & set paths.